### DISCUSSION TITLE
Vaultwarden + Traefik + Custom error pages middleware broke 2FA

### DISCUSSION THREAD

**Discussion by Uzurka** [dani-garcia/vaultwarden#3873] [Q&A]
### Subject of the issue
I'm using Vaultwarden for a moment now, without any problem, using traefik as reverse proxy.
Recently, i've added to my traefik configuration a service & middleware for customise error-pages (404, 500, 501, and all other errors)
I've added this middleware to my Vaultwarden container, resulting 2FA code is not prompted anymore.
Instead, i get this error : 
[2023-09-08 21:12:29.267][error][ERROR] 2FA token not provided

And on the web, a red rectangle with an unknown error

### Your environment (Generated via diagnostics page)

* Vaultwarden version: v1.29.2
* Web-vault version: v2023.7.1
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: MySQL
* Database version: 10.11.5-MariaDB-1:10.11.5+maria~ubu2204
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, SIGNUPS_ALLOWED, INVITATIONS_ALLOWED, SHOW_PASSWORD_HINT, ADMIN_TOKEN, INVITATION_ORG_NAME, IP_HEADER, SMTP_HOST, SMTP_SECURITY, SMTP_PORT, SMTP_FROM, SMTP_FROM_NAME, SMTP_USERNAME, SMTP_PASSWORD

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "*****://******************************************************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*******************",
  "domain_origin": "*****://*******************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": "api-02b38f68.duosecurity.com",
  "duo_ikey": "DIF8SGQ3TOV9LBLRJ8XG",
  "duo_skey": "***",
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Bitwarden - Uzurka",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "warn",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": false,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "*******************",
  "smtp_from_name": "Bitwarden - Uzurka",
  "smtp_host": "**************",
  "smtp_password": "***",
  "smtp_port": 465,
  "smtp_security": "force_tls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "*******************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

* Install method: Docker container deployed with Ansible

* Reverse proxy and version: Traefik v2.10.4

* MySQL/MariaDB or PostgreSQL version: MariaDB latest

* Other relevant details:

### Steps to reproduce
Add a middleware to the container that replaces error-pages with custom ones

### Expected behaviour
2FA code is prompted

### Actual behaviour
2FA code isn't prompted

**Comment by BlackDex** [comment]
Well you figured it out your self already what is wrong, it's the middleware. Prevent any 4xx or 5xx message to be altered for Vaultwarden and it is fixed.

The clients expect specific json messages to take the correct action. The middleware is preventing this by replacing these messages.

**Comment by mrdavidkovacs** [comment]
I know this is a old discussion, but there are many other issues/discussion which describes the same/similar problem: #6636, #6606, #6333.
It should not be necessary to disable a middleware/WAF in order to get the luxury of 2FA.

**Reply by BlackDex** [reply]
You can always add exclusions for those specific endpoints, but there are many other aspects with waf/security middleware which might cause issues with the clients if those modify headers or content.

**Reply by mrdavidkovacs** [reply]
Currently, traefik does not support disabling entry point level middlewares on a router level.
This applies to all traefik middlewares.

So one could say, that traefik should support that, and yes, I agree, but on the other hand I don't think that it is a good idea to rely on perfect world assumptions.
And there will be rewrites of responses with error codes.
If it is the assumed workflow the API should operate with 2XX status codes.

**Reply by BlackDex** [reply]
Then please report an issue at Bitwarden, but that will probably fail as you use Vaultwarden.
You could try reporting it by using Bitwarden Lite, which is a self-hosted instance too, but not sure if they support middlewares or WAF which alter responses, not sure if there are reports of this already btw, that might also be the case.

Also, from my point of view, a WAF/Security Middleware should never alter API call responses, unless they really need to block something, And if it blocks something which is a valid request, than that should be excluded or fixed by those tools.

It's not up to this project to change the clients behavior.
We have added several checks to at least try and check for broken responses because of these security tools to help prevent issues or at least detect them quickly.

There is nothing on the server side we can fix to let those WAF's or middlewares which alter responses still let the clients show the right responses.

**Reply by mrdavidkovacs** [reply]
Thank you very much.
You're right, this repo does not contain any client.
For such a change we would need to change both parts: client and server.
